src/sb-util.ts

```diff
@@ -1,32 +1,215 @@
+
+		return new SpriteCollection(sprites).query(selector);
+	}
+
+	stage() {
+		return this.sprites('[isStage=true]');
+	}
+
+	blocks() {
+		let blocks = [];
+		let sprites = this.sprites();
+		for (let s of sprites) {
+			const blocksObj = s.prop(SpriteAttributes.BLOCKS);
+			Object.entries(blocksObj).forEach(([key, val]) => {
+				let block = val;
+				block[BlockAttributes.ID] = key;
+				blocks.push(block);
+			})
+		}
+		return new BlockCollection(blocks);
 	}
 }
 
-export const loadSb3 = function(source: string) {
-	return new Promise<ScratchProject>(async (resolve, reject) => {
+export class SpriteCollection implements Queryable  {
+	constructor(sprites: Array<SpriteProperties>){
+		storage.set(this, sprites);
+	}
+
+	prop(attribute: string) {
+		const first = storage.get(this).slice().shift();
+		return first[attribute];
+	}
+
+	/*
+		Currently, the query selector syntax only supports
+		attribute selector style in the form of [attr] and
+		[attr=value]
+	*/
+	query(selector: string){
+		if(selector.length < 2 || selector[0] !== SelectorSyntax.OPEN_BRACKET
+			&& selector.slice(-1) !== SelectorSyntax.CLOSED_BRACKET){
+			throw new Error('Invalid selector syntax for SpriteCollection. \
+				[attr] and [attr=value] format is accepted. Selector must not be empty');
+		}
+
+		// string between brackets
+		const selectorBody = selector.substring(1, selector.length-1);
+
+		let sprites, allSprites = storage.get(this);
+
+		// case when selector string is in [attr=value] form
+		if(selectorBody.includes(SelectorSyntax.EQUALS)) {
+			const [attr, valueString] = selectorBody.split(SelectorSyntax.EQUALS);
+
+			this['value'] = valueString;
+
+			// handle case when booleans are strings
+			if (valueString === 'true' || valueString === 'false') {
+				this['value'] = Boolean(valueString);
+			}
+			// handle case when numbers are strings
+			else if (!isNaN(+valueString)) {
+				this['value'] = +valueString;
+			}
+			// handle case when strings have quotes
+			else if ((valueString[0] === SelectorSyntax.DOUBLE_QUOTE || valueString[0] === SelectorSyntax.SINGLE_QUOTE)
+					&& (valueString.slice(-1) === SelectorSyntax.DOUBLE_QUOTE || valueString.slice(-1) === SelectorSyntax.SINGLE_QUOTE)) {
+				this['value'] = valueString.replace(/^"(.*)"$/, '$1');
+			}
+
+			sprites = allSprites.filter((s: SpriteProperties) => s[attr] === this['value']);
+		}
+		// case when selector string is in [attr] form
+		else {
+			const attr = selectorBody;
+			sprites = allSprites.filter((s: SpriteProperties) => attr in s);
+		}
+
+		return sprites.length === 1 ? new Sprite(sprites.pop()) : new SpriteCollection(sprites);
+	}
+
+	[Symbol.iterator]() {
+		const sprites = storage.get(this);
+		return {
+			next(): IteratorResult<Sprite> {
+				let sprite = sprites.pop();
+				if(sprite){
+					return {value: new Sprite(sprite), done: false};
+				}
+				return {done: true, value: null};
+			}
+		}
+	}
+}
+
+export class Sprite extends SpriteCollection implements Queryable {
```

Comment:
not sure you even need this class you can just pass a `SpriteCollection` with a length of 1 instead?